matrix/matrix.py

```diff
@@ -0,0 +1,146 @@
+"""Matrix data type class."""
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type.
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+            the matrix is represented by a tuple.
+    note:
+        This implementation only supports squared matrices (n * n).
+    """
+
+    def __init__(self, lines):
+        self._validate(lines)
+        self._matrix = lines
+
+    def _validate(self, data: tuple):
+        """Check if the data given is valid.
+
+        note:
+            Valid data is defined if it contains tuples of the same length
+            (Matrix only supports n * n), items can't be iterable.
+        """
+        # if any(hasattr(item, '__len__') for line in data for item in line):
+        #     raise TypeError("Matrix item type not supported")
+
+        if not all([len(line) == len(data) for line in data]):
+            raise ValueError("Matrix must be even-sided")
+
+    @classmethod
+    def unity(cls, dimension: int):
+        """Generate a matrix with a main diagonal of 1's, elsewhere 0's."""
+        return cls(
+            tuple(tuple((0,) * i + (1,) + (0,) * (dimension - i - 1))
+                  for i in range(dimension)))
+
+    @classmethod
+    def ones(cls, dimension):
+        """Generate a matrix full of 1's."""
+        return cls(((1,) * dimension,) * dimension)
+
+    def __repr__(self):
+        return f"'{self.__class__.__name__}({self._matrix})'"
+
+    def __str__(self):
+        """Return a string presentation of the matrix instance."""
+        return repr(self._matrix)
+
+    @property
+    def tuples(self):
+        """Return the matrix instance as a tuple of tuples."""
+        return self._matrix
+
+    def _add_matrix(self, other):
+        """Add two matrices.
+
+        Args:
+            other (Matrix): the other matrix to operate on.
+
+        Returns:
+            Matrix: the result of the addition.
+        """
+        dimension = len(self.tuples)
+
+        if len(other.tuples) != dimension:
+            raise ValueError("Matrices length don't match")
+
+        calculated_data = [[0 for i in range(dimension)]
+                           for j in range(dimension)]
+
+        for i in range(len(self.tuples)):
+            for j in range(len(self.tuples)):
+                calculated_data[i][j] = self.tuples[i][j] + other.tuples[i][j]
+
+        calculated_data = [tuple(line) for line in calculated_data]
+        return Matrix(tuple(calculated_data))
+
+    def _matrix_mul_matrix(self, other):
```

Comment:
You do the exact same validation for multiple function... try to read about decorators